veps/sig-api/kubectl_vm_graceful_deletion.md

```diff
@@ -0,0 +1,129 @@
+- **Developers**: KubeVirt developers require rapid VM cleanup for iterative development or automated testing.
+
+## User Stories
+
+- *As a VM Owner*, I want to delete a VM with a custom grace period using `kubectl delete`, avoiding delays from long default grace periods.
+- *As a developer*, I want a single `kubectl` command to terminate VMs quickly without relying on `virtctl`.
+
+## Repositories Affected
+
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+
+## Design
+
+A **validating webhook** will be implemented to handle `DELETE` operations for `VirtualMachine` resources.
+validating and applying the user-specified `--grace-period` by directly patching the associated `VirtualMachineInstance` (VMI).
+
+### Workflow
+
+1. **Interception and Validation**: The validating webhook intercepts
+   `DELETE` API calls targeting `VirtualMachine` resources. If the user
+   provides a `--grace-period=X` value, the webhook validates that it is a
+   non-negative integer.
+2. **VMI Patching**: If valid, the webhook patches the corresponding VMI 
+   `spec.terminationGracePeriodSeconds` to the user-specified value.
+3**Error Handling**: If the patch fails, the webhook rejects the
+   deletion request with a descriptive error.
+4**Deletion Propagation**: The VM deletion proceeds as usual. The updated
+   VMI grace period governs how the launcher pod shuts down.
+
+### Pros
+
+- **User-friendly**: Aligns `kubectl delete` behavior with Kubernetes standards.
+- **Backward-compatible**: Users not using `--grace-period` experience no change.
+
+### Cons
+
+- **Custom logic**: Requires KubeVirt-specific webhook to handle a standard flag.
+
+## API Usage Example
+
+```bash
+kubectl delete vm myvm --grace-period=5
+```
+
+This command deletes the VM `myvm` with a 5-second grace period, overriding any longer default period.
+
+## Alternatives
+
+### Option 1: Enhance Kubernetes Core - already tried this approach 8 years ago
+
+enhance Kubernetes so that the `--grace-period` flag from `kubectl delete` is propagated to a resource’s `metadata.terminationGracePeriodSeconds`, even for custom resources like `VirtualMachine`.
+today when custom resources get deleted, the value for `metadata.terminationGracePeriodSeconds` get populated to 0 regardless of the flag given by the user.
+
+This path was already explored in kubernetes/kubernetes#56567 about 8 years ago, but was never fully implemented.
+That issue aimed to allow `DELETE` calls to honor the --grace-period flag for all resources—including custom ones—but eventually stalled due to complexity and lack of prioritization.
+
+### Pros
+
+- **Generic solution**: Works for all custom resources, not just VMs.
+- **Standard behavior**: Eliminates the need for custom webhook logic.
+
+### Cons
+
+- **Complex coordination**: Requires changes across multiple Kubernetes SIGs.
+
+## Scalability
+No scalability concerns are anticipated, as the webhook operates on individual `DELETE` operations.
```

Comment:
I think scalability is something to consider here.
If we're going to use a webhook, it means all load is on virt-api (which has [scalability issues](https://github.com/kubevirt/kubevirt/issues/7101) of its own...). Think about mass-deletion of thousands of VMs at once. This will definitely increase the load on virt-api.

I'm not saying it's an issue, but definitely something to explore.